### PR TITLE
Add maxAge field to MachineHealthChecks

### PR DESCRIPTION
**Pull request by alexander-demichev**
Add maxAge field to MachineHealthChecks. If exists more than maximal age allowed it is marked for remediation.

### CODE REVIEW

File: install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml [hunk 1]
```diff
@@ -50,6 +50,10 @@ spec:
           spec:
             description: Specification of machine health check policy
             properties:
+              maxAge:
+                description: MaxAge maximal age of a machine before it gets remediated Expects an unsigned duration string of decimal numbers each with optional fraction and a unit suffix, eg "300ms", "1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
```
Comment (elmiko) [inline install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml:54]: looks like a period is missing
```suggestion
                description: MaxAge maximal age of a machine before it gets remediated. Expects an unsigned duration string of decimal numbers each with optional fraction and a unit suffix, eg "300ms", "1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
```

File: pkg/apis/machine/v1beta1/machinehealthcheck_types.go [hunk 2]
```diff
@@ -80,6 +80,15 @@ type MachineHealthCheckSpec struct {
...
 	NodeStartupTimeout metav1.Duration `json:"nodeStartupTimeout,omitempty"`
+
+	// MaxAge maximal age of a machine before it gets remediated
+	// Expects an unsigned duration string of decimal numbers each with optional
+	// fraction and a unit suffix, eg "300ms", "1.5h" or "2h45m".
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// +optional
+	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	MaxAge *metav1.Duration `json:"maxAge"`
```
Comment (JoelSpeed) [inline pkg/apis/machine/v1beta1/machinehealthcheck_types.go:91]: Does this need to be a pointer? Could we just use zero as the disabled state? Could use omitEmpty so it doesn't show up when it is set to zero

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 3]
```diff
@@ -1830,6 +1830,59 @@ func TestNeedsRemediation(t *testing.T) {
...
+			target: &target{
+				Machine: mapiv1beta1.Machine{
+					TypeMeta: metav1.TypeMeta{Kind: "Machine"},
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations:       make(map[string]string),
+						Name:              "machine",
+						Namespace:         namespace,
+						Labels:            map[string]string{"foo": "bar"},
+						OwnerReferences:   []metav1.OwnerReference{{Kind: "MachineSet"}},
+						CreationTimestamp: metav1.Time{Time: time.Now().AddDate(0, 0, -1)},
```
Comment (JoelSpeed) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go:1844]: This isn't immediately obvious what this does, does it make it 1 day ago? 
```suggestion
						CreationTimestamp: metav1.Time{Time: time.Now().AddDate(0, 0, -1)}, // Now minus one day
```

**Review by elmiko — COMMENTED** [review]
looks mostly good to me, one minor nit

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 4]
```diff
@@ -1830,6 +1830,59 @@ func TestNeedsRemediation(t *testing.T) {
...
+					TypeMeta: metav1.TypeMeta{
+						Kind: "MachineHealthCheck",
+					},
+					Spec: mapiv1beta1.MachineHealthCheckSpec{
+						Selector: metav1.LabelSelector{
+							MatchLabels: map[string]string{
+								"foo": "bar",
+							},
+						},
+						MaxAge: &metav1.Duration{Duration: time.Second},
```
Comment (JoelSpeed) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go:1876]: If we can make the machine a day old, we should make this like 20 hours or something, so it's not just any machine that was created. Being a large time difference would make it more obvious what the test is doing rather than a time period that could elapse during the duration of the test.
```suggestion
						MaxAge: &metav1.Duration{Duration: time.Second},
```

Should also add a test case where it hasn't reached the maximum age

File: pkg/apis/machine/v1beta1/machinehealthcheck_types.go [hunk 5]
```diff
@@ -80,6 +80,15 @@ type MachineHealthCheckSpec struct {
 	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
 	// +kubebuilder:validation:Type:=string
 	NodeStartupTimeout metav1.Duration `json:"nodeStartupTimeout,omitempty"`
+
+	// MaxAge maximal age of a machine before it gets remediated
```
Comment (JoelSpeed) [inline pkg/apis/machine/v1beta1/machinehealthcheck_types.go:84]: Let's expand this to be a bit more about the use case
```suggestion
	// Machines older than this duration will be considered to have reached their maximal age and will be remediated.
	// This can be used to ensure that Machines are replaced frequently.
```

**Review by michaelgugino — CHANGES_REQUESTED** [review]
We should compare instance provisioning time rather than machine creation timestamp as the two may be significantly different, up to 2 hours today.  In the future, I would like the 2 hour window for CSR approval to be based on instance creation time, this will allow the usecase of creating a bunch of machines that spin until price/capacity is available for spot instances.

Also, the field name maxAge is not specific.  We should use something like maxInstanceAge

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 3 above)
Comment (alexander-demichev) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go:1844]: yes, I'll add the comment

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 4 above)
Comment (alexander-demichev) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: I added a new test, thanks

File: pkg/apis/machine/v1beta1/machinehealthcheck_types.go (hunk 2 above)
Comment (alexander-demichev) [reply pkg/apis/machine/v1beta1/machinehealthcheck_types.go]: I added omitEmpty

**Comment by enxebre** [comment]
How will we prevent all machines in a scalable resource which might have very similar creation timestamp from being deleted at a time? So we don't drain multiple times relocating workload inefficiently?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 6]
```diff
@@ -638,6 +638,14 @@ func (t *target) needsRemediation(timeoutForMachineToHaveNode time.Duration) (bo
 		return false, nextCheck, nil
 	}
 
+	if t.MHC.Spec.MaxAge != nil {
+		machineCreationTime := t.Machine.CreationTimestamp.Time
+		if machineCreationTime.Add(t.MHC.Spec.MaxAge.Duration).Before(now) {
+			klog.V(3).Infof("%s: unhealthy: machine reached maximal age %v", t.string(), t.MHC.Spec.MaxAge.Duration)
```
Comment (seanmalloy) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:644]: Beyond just logging a message does the MHC generate k8s events or have prometheus metrics?

We have some similar-ish node killer automation running in our OCP v3.11 environment and we found it very useful to create k8s events and have prometheus metrics for monitoring/tracking old nodes that are being deleted in our production environment.

**Comment by seanmalloy** [comment]
> How will we prevent all machines in a scalable resource which might have very similar creation timestamp from being deleted at a time? So we don't drain multiple times relocating workload inefficiently?

Could this be successfully mitigated by having the user also create a [MachineDisruptionBudget](https://github.com/openshift/machine-api-operator/blob/master/docs/proposals/machine-disruption-budget-kpe.md)?

In our real word OCP v3.11 cluster we delete one "old node" every two hours. I suppose even with an MDB the MHC could delete a lot more than one machine every two hours.
